sklearn/tree/_tree.pyx

```diff
@@ -443,44 +472,72 @@ cdef class ClassificationCriterion(Criterion):
             The new ending position for which to move samples from the right
             child to the left child.
         """
-
         cdef DOUBLE_t* y = self.y
         cdef SIZE_t y_stride = self.y_stride
         cdef DOUBLE_t* sample_weight = self.sample_weight
 
         cdef SIZE_t* samples = self.samples
         cdef SIZE_t pos = self.pos
+        cdef SIZE_t end = self.end
 
         cdef SIZE_t n_outputs = self.n_outputs
         cdef SIZE_t* n_classes = self.n_classes
         cdef SIZE_t label_count_stride = self.label_count_stride
         cdef double* label_count_left = self.label_count_left
         cdef double* label_count_right = self.label_count_right
+        cdef double* label_count_total = self.label_count_total
 
-        cdef SIZE_t i
-        cdef SIZE_t p
-        cdef SIZE_t k
-        cdef SIZE_t label_index
+        cdef SIZE_t i = 0
+        cdef SIZE_t p = 0
+        cdef SIZE_t k = 0
+        cdef SIZE_t c = 0
+        cdef SIZE_t label_index = 0
         cdef DOUBLE_t w = 1.0
         cdef DOUBLE_t diff_w = 0.0
 
-        # Note: We assume start <= pos < new_pos <= end
         # Go through each sample in the new indexing
-        for p in range(pos, new_pos):
-            i = samples[p]
+        if (new_pos - pos) <= (end - new_pos):
```

Comment:
Could you add a comment here explaining the motivation of this trick? Why would it be more efficient to count backwards?